### PR TITLE
[core] Reduce lock contention in TUrl::GetSpecialProtocols().

### PR DESCRIPTION
**Pull request by Axel-Naumann**
Fixes MT overhead of TChain::Add

### CODE REVIEW

File: core/base/src/TUrl.cxx [hunk 1]
```diff
@@ -584,14 +586,15 @@ TObjArray *TUrl::GetSpecialProtocols()
...
+   R__LOCKGUARD(gROOTMutex);
    if (fgSpecialProtocols)
       fgSpecialProtocols->Delete();
 
    if (!fgSpecialProtocols)
       fgSpecialProtocols = new TObjArray;
 
    const char *protos = gEnv->GetValue("Url.Special", "file: hpss: dcache: dcap:");
-   usedEnv = kTRUE;
+   usedEnv = true;
```
Comment (pcanal) [inline core/base/src/TUrl.cxx:597]: Need to be move to the very end to avoid a 2nd thread used partially initialized state.

Comment (Axel-Naumann) [reply core/base/src/TUrl.cxx:597]: Doesn't the `R__LOCKGUARD(gROOTMutex);` a few lines above (line 589 in the new file) prevent that?

**Review by eguiraud — APPROVED** [review]
looks good to me!

**Comment by Axel-Naumann** [comment]
...assuming the i386 failure is spurious...

**Comment by pcanal** [comment]
No, if both thread arrive at the lock at the same time (with usedEnv==false).

On 11/23/20 2:08 AM, Axel Naumann wrote:

@Axel-Naumann commented on this pull request.

________________________________

In core/base/src/TUrl.cxx<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_root-2Dproject_root_pull_6857-23discussion-5Fr528523942&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=UWfwI6PGOKY1hmtnuMh4_A&m=ukfUznEXjXBdxDymQ6EAmLvJOoEkGLJ5U7MGPdhKPFA&s=co6lV5rF4ChE-w-mEfaE-93x95PezBrxdUj5efG1DSQ&e=>:

>     if (fgSpecialProtocols)
       fgSpecialProtocols->Delete();

    if (!fgSpecialProtocols)
       fgSpecialProtocols = new TObjArray;

    const char *protos = gEnv->GetValue("Url.Special", "file: hpss: dcache: dcap:");
-   usedEnv = kTRUE;
+   usedEnv = true;


Doesn't the R__LOCKGUARD(gROOTMutex); a few lines above (line 589 in the new file) prevent that?

**Comment by Axel-Naumann** [comment]
Indeed, which is why I removed my wrong comment and fixed the code! :-)

**Pull request merged**